keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2314 @@
+        ...
+        // Used is the current observed total usage of the resource in the namespace
+        // +optional
+        Used ResourceList
++       // QOSResources contains the enforced set of available QoS resources.
++       // +optional
++       QOSResources QOSResourceQuota
++       // QOSResourcesUsage contains the observed usage of QoS resources in the namespace.
++       // +optional
++       QOSResourcesUsage QOSResourceQuota
+ }
+```
+
+#### Scheduler improvements
+
+The first implementation phase only adds basic filtering of nodes based on
+QoS-class resources and node scoring is not altered. However, the relevant
+scheduler plugins (e.g. NodeResourcesFit and NodeResourcesBalancedAllocation)
+could be extended to do scoring based on the capacity, availability and usage
+of QoS-class resources on the nodes.
+
+#### Kubelet-initiated pod eviction
+
+QoS-class resources available on a node are dynamic in the sense that they may
+change over the lifetime of the node. E.g. re-configuration of the container
+runtime may make new types of QoS-class resources available, properties of
+existing resources may changes (e.g. the set of available classes or their
+capacity) or some resources might be removed completely. It might be desirable
+that kubelet could evict a running pod that request QoS-class resources that
+are no more available on the node. This should be relatively straightforward to
+implement as kubelet knows what QoS-class resources are available on the node
+and also monitors all running pods.
+
+#### Default and limits
+
+LimitRanges is a way to specify per-container resource constraints. They might
+have limited applicability also in the context of QoS-class resources. In
+particular, LimitRanges could be used for specifying defaults for QoS-class
+resources. An additional possible usage would be to set per-pod limits on the
+usage of container-level QoS-class resources.
+
+```diff
+ // LimitRangeItem defines a min/max usage limit for any resource that matches on kind
+ type LimitRangeItem struct {
+@@ -5167,6 +5167,28 @@ type LimitRangeItem struct {
+        // MaxLimitRequestRatio represents the max burst value for the named resource
+        // +optional
+        MaxLimitRequestRatio ResourceList
++       //  QOSResources specifies the limits for QoS resources.
++       QOSResources []LimitQOSResource
++}
++
++// LimitQOSResource specifies limits of one QoS resources type.
++type LimitQOSResource struct {
++       // Name of the resource.
++       Name QOSResourceName
++       // Default specifies the default class to be assigned.
++       // +optional
++       Default string
++       // Max usage of classes
++       // +optional
++       Max []QOSResourceClassLimit
++}
++
++// QOSResourceClassLimit specifies a limit for one class of a QoS resource.
++type QOSResourceClassLimit struct {
++       // Name of the class.
++       Name string
++       // Capacity is the limit for usage of the class.
++       Capacity int64
```

Comment:
nit: I prefer `Ceiling`

Reply:
Noted. I added this as UNRESOLVED in the KEP. Just a reminder that all the stuff under `Future work` is unresolved for the time being but made this detail explicit...